clang/lib/Headers/hlsl/hlsl_intrinsics.h

```diff
@@ -80,6 +80,25 @@ void asuint(double3, out uint3, out uint3);
 _HLSL_BUILTIN_ALIAS(__builtin_hlsl_elementwise_splitdouble)
 void asuint(double4, out uint4, out uint4);
 
+//===----------------------------------------------------------------------===//
+// asint16 builtins
+//===----------------------------------------------------------------------===//
+
+/// \fn int16_t asint16(T Val)
+/// \brief Interprets the bit pattern of x as an 16-bit integer.
+/// \param Val The input value.
+#ifdef __HLSL_ENABLE_16_BIT
+
+template <typename T, int N>
+constexpr vector<int16_t, N> asint16(vector<T, N> V) {
+  return __detail::bit_cast<int16_t, T, N>(V);
+}
+
+template <typename T> constexpr int16_t asint16(T F) {
```

Comment:
This isn't quite correct correct. T here needs to be constrained for both the scalar and vector cases. The docs say asint16 is only valid for [half](https://github.com/microsoft/DirectXShaderCompiler/wiki/16-Bit-Scalar-Types), [int16_t](https://github.com/microsoft/DirectXShaderCompiler/wiki/16-Bit-Scalar-Types), or [uint16_t](https://github.com/microsoft/DirectXShaderCompiler/wiki/16-Bit-Scalar-Types).

you will need to do something like 
```c++
enable_if_t<is_same< int16_t, T>::value || is_same< uint16_t, T>::value || is_same<half, T>::value, T>
```

There should be examples  you can find in the headers that do similar things you can extrapolate a solution from.

Reply:
Ok so after comfirming with other folks we should constrain the supported overloads of `asint16` to `half`, `int16_t`, and `uint16_t`. Using the `enable_if_t` strategy indicated above. This will impact the tests in `clang/test/SemaHLSL/BuiltIns/asint16-errors.hlsl` to all be `// expected-error@-1 {{no matching function for call to 'asint16'}}` for all the non 16 bit cases.